src/AutoRest.CSharp/Common/Output/Models/Types/SignatureType.cs

```diff
@@ -0,0 +1,318 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using AutoRest.CSharp.Common.Input;
+using AutoRest.CSharp.Common.Output.Expressions.Statements;
+using AutoRest.CSharp.Common.Output.Expressions.ValueExpressions;
+using AutoRest.CSharp.Common.Output.Models;
+using AutoRest.CSharp.Generation.Types;
+using AutoRest.CSharp.Input.Source;
+using AutoRest.CSharp.Output.Models.Shared;
+using Microsoft.CodeAnalysis;
+using static AutoRest.CSharp.Common.Output.Models.Snippets;
+
+namespace AutoRest.CSharp.Output.Models.Types
+{
+    /// <summary>
+    /// This type holds three portions of codes:
+    ///     - current
+    ///     - custom
+    ///     - baseline contract
+    ///     current union custom compare with baseline contract outputs the changeset, we can apply different rules with it.
+    /// </summary>
+    internal class SignatureType
+    {
+        private readonly TypeFactory _typeFactory;
+        private readonly string _defaultNamespace;
+        private readonly string _defaultName;
+        private readonly SignatureType? _customization;
+        private readonly SignatureType? _baselineContract;
+        private readonly MethodChangeset? _methodChangeset;
+
+        // Missing means the method with the same name is missing from the current contract
+        // Updated means the method with the same name is updated in the current contract, and the list contains the previous method and current methods including overload ones
+        private record MethodChangeset(IReadOnlyList<MethodSignature> Missing, IReadOnlyList<(List<MethodSignature> Current, MethodSignature Previous)> Updated) { }
+
+        public SignatureType(TypeFactory typeFactory, IReadOnlyList<MethodSignature> methods, SourceInputModel? sourceInputModel, string defaultNamespace, string defaultName)
```

Comment:
I believe we should remove the word default here (see the comment in src/AutoRest.CSharp/Common/Output/Models/Types/ModelFactoryTypeProvider.cs)